syntax = "proto2";

message InputState {
    required bool   accelerate = 1;
    required bool   brake      = 2;
    required bool   item       = 3;
    required bool   drift      = 4;
    required bool   brakeDrift = 5;
    required uint32 stickX     = 6;
    required uint32 stickY     = 7;
    required uint32 trick      = 8;
}

message PlayerFrame {
    message Vec3 {
        required float x = 1;
        required float y = 2;
        required float z = 3;
    }

    message Quat {
        required float x = 1;
        required float y = 2;
        required float z = 3;
        required float w = 4;
    }

    required InputState inputState          = 1;
    required uint32     timeBeforeRespawn   = 2;
    required uint32     timeInRespawn       = 3;
    repeated uint32     timesBeforeBoostEnd = 4;
    required Vec3       pos                 = 5;
    required Quat       mainRot             = 6;
    required float      internalSpeed       = 7;
}

message RaceClientPing {}

message RaceClientFrame {
    required uint32      time       = 1;
    required uint32      serverTime = 2;
    repeated PlayerFrame players    = 3;
}

message RaceServerFrame {
    required uint32      time        = 1;
    repeated uint32      playerTimes = 2;
    repeated PlayerFrame players     = 3;
}
